lib/src/yaml_edit/editor.dart

```diff
@@ -0,0 +1,679 @@
+          windowsNewlines++;
+        } else {
+          unixNewlines++;
+        }
+      }
+      _isWindowsEnding = windowsNewlines > unixNewlines;
+    }
+
+    return _isWindowsEnding ? '\r\n' : '\n';
+  }
+
+  bool _isWindowsEnding;
+
+  factory YamlEditor(String yaml) => YamlEditor._(yaml);
+
+  YamlEditor._(this._yaml) {
+    ArgumentError.checkNotNull(_yaml);
+    _initialize();
+  }
+
+  void _initialize() {
+    _contents = loadYamlNode(_yaml);
+    _aliases = {};
+
+    var stack = [_contents];
+    var encounteredNodes = <YamlNode>{};
+
+    while (stack.isNotEmpty) {
+      var nextElement = stack.removeLast();
+
+      if (!encounteredNodes.contains(nextElement)) {
+        encounteredNodes.add(nextElement);
+      } else {
+        _aliases.add(nextElement);
+      }
+
+      if (nextElement is YamlMap) {
+        stack.addAll(nextElement.nodes.values);
+      } else if (nextElement is YamlList) {
+        stack.addAll(nextElement.nodes);
+      }
+    }
+
+    _isWindowsEnding = null;
+    _indentation = null;
+  }
+
+  /// Parses the document to return [YamlNode] currently present at [path].
+  ///
+  /// If no [YamlNode]s exist at [path], the result of invoking the [orElse]
+  /// function is returned.
+  ///
+  /// If [orElse] is omitted, it defaults to throwing a [PathError].
+  ///
+  /// To get `null` when [path] does not point to a value in the [YamlNode]-tree,
+  /// simply pass `orElse: () => null`.
+  /// ```dart
+  /// final myYamlEditor('{"key": "value"}');
+  /// final value = myYamlEditor.valueAt(['invalid', 'path'], orElse: () => null);
+  /// print(value) // null
+  /// ```
+  ///
+  /// Common usage example:
+  /// ```dart
+  ///   final doc = YamlEditor('''
+  /// a: 1
+  /// b:
+  ///   d: 4
+  ///   e: [5, 6, 7]
+  /// c: 3
+  /// ''');
+  /// print(doc.parseAt(['b', 'e', 2])); // 7
+  /// ```
+  ///
+  /// To illustrate that [parseAt] returns a [YamlNode] that can be invalidated
+  /// if modifications are performed to the document,
+  /// ```dart
+  /// final doc = YamlEditor("YAML: YAML Ain't Markup Language");
+  /// final node = doc.parseAt(['YAML']);
+  ///
+  /// print(node.value); // Expected output: "YAML Ain't Markup Language"
+  ///
+  /// doc.update(['YAML'], 'YAML');
+  ///
+  /// final newNode = doc.parseAt(['YAML']);
+  ///
+  /// // Note that the value does not change
+  /// print(newNode.value); // "YAML"
+  /// print(node.value); // "YAML Ain't Markup Language"
+  /// ```
+  YamlNode parseAt(Iterable<Object> path, {YamlNode Function() orElse}) {
+    ArgumentError.checkNotNull(path, 'path');
+
+    return _traverse(path, orElse: orElse);
+  }
+
+  /// Sets [value] in the [path].
+  ///
+  /// Note that [update] provides a different result than a [remove] followed by
+  /// an [insertIntoList], because it preserves comments at the same level.
+  ///
+  /// Throws a [PathError] if [path] is invalid.
+  ///
+  /// ```dart
+  /// final doc = YamlEditor('''
```

Comment:
```suggestion
  /// **Example:** (using [update])
  /// ```dart
  /// final doc = YamlEditor('''
```